api/python/quilt3/data_transfer.py

```diff
@@ -171,6 +178,61 @@ def _build_unsigned_client(self):
         s3_client = self._build_client(lambda session: True)
         self._unsigned_client = s3_client
 
+    @staticmethod
+    def add_options_safely(params: dict, options: Optional[dict]):
+        """
+        Add options to the params dictionary safely.
+        This method ensures that the options do not overwrite existing keys in the params dictionary.
+        """
+        if options:
+            for key, value in options.items():
+                if key in params:
+                    raise ValueError(f"Cannot override key `{key}` using options: {options}.")
+                params[key] = value
+
+    def register_event_callback(self, event_name: str, callback: Optional[Callable]) -> None:
+        """Register a callback for S3 client events.
+
+        Args:
+            event_name: The name of the event to register for (e.g. 'creating-client-class')
+            callback: The callback function to be called when the event occurs, or None to remove the callback
+        """
+        if callback is None:
+            del self._event_callbacks[event_name]
+        else:
+            self._event_callbacks[event_name] = callback
```

Comment:
might make sense to not fail when handler being removed doesn't exist (or fail when you add duplicate event handler)

BTW boto3 supports multiple handlers for the same event
we *might want* to support that as well though we don't have to

Reply:
My thought was that the easiest way to remove a callback is to set it to None.  But I guess it is effectively reset after every call, so it doesn't matter.